### PR TITLE
Make `FilteringTextInputFormatter`'s filtering Selection/Composing Region agnostic

### LINKED ISSUE #80842: `FilteringTextInputFormatter`'s output text is selection dependent
The following test fails:

```dart
  test('FilteringTextInputFormatter should filter independent of selection', () {
    final TextInputFormatter formatter = FilteringTextInputFormatter.deny('abc', replacementString: '*');

    const TextEditingValue oldValue = TextEditingValue();
    const TextEditingValue newValue = TextEditingValue(text: 'abcabcabc');

    final String filteredText = formatter.formatEditUpdate(oldValue, newValue).text;

    for (int i = 0; i < newValue.text.length; i += 1) {
      final String text = formatter.formatEditUpdate(
        oldValue,
        newValue.copyWith(selection: TextSelection.collapsed(offset: i)),
      ).text;
      expect(filteredText, text);
    }
  });
```

because currently the text is not formatted as a single string:

https://github.com/flutter/flutter/blob/f2417e6c1cc9514c2568187e5b4082475872a304/packages/flutter/lib/src/services/text_formatter.dart#L542-L551

### LINKED ISSUE #87764: Document the composing text feature
The [composing](https://master-api.flutter.dev/flutter/services/TextEditingValue/composing.html) feature of Flutter's text editing doesn't seem to be explained anywhere in the docs.  The docs linked above simply state:

```
The range of text that is still being composed.
```

What this feature does is to support the IME feature that allows replacing the text currently being edited with some replacement.  This is used, for example, by non-phonetic writing systems:

Screen Shot 2021-08-05 at 4 58 19 PM

And even by Gboard to do suggestions (note the underline in the field and suggestions in the keyboard):

We should explain this in the docs!

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
- Fixes https://github.com/flutter/flutter/issues/80842
- Fixes https://github.com/flutter/flutter/issues/87764

Removed the additional interfaces added from the original (closed) PR.

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/services/text_input.dart [hunk 1]
```diff
@@ -722,9 +722,38 @@ class TextEditingValue {
...
   /// The range of text that is currently selected.
+  ///
+  /// When [selection] is a [TextSelection] that has the same non-negative
+  /// `baseOffset` and `extentOffset`, the [selection] property represents the
+  /// caret position.
+  ///
+  /// If the current [selection] has a negative `baseOffset` or `extentOffset`,
+  /// then the text currently does not have a selection or a caret location, and
+  /// most text editing operations that rely on the current selection (for
+  /// instance, insert a character at the caret location) will do nothing.
```
Comment (justinmc) [inline packages/flutter/lib/src/services/text_input.dart:733]: Thank you for documenting these quirks 🙏

File: packages/flutter/lib/src/services/text_input.dart [hunk 2]
```diff
@@ -722,9 +722,38 @@ class TextEditingValue {
...
+  /// enter a latin alphabet sequence and then convert it to CJK characters. On
+  /// iOS, the default software keyboards do not have a dedicated view to show
+  /// the unfinished latin sequence, so it's displayed directly in the text
+  /// field, inside of a composing region.
+  ///
+  /// The composing region should typically only be changed by the IME, or the
+  /// user via interacting with the IME.
+  ///
+  /// If the range represented by this property is [TextRange.empty], then the
+  /// text is not currently being composed.
```
Comment (justinmc) [inline packages/flutter/lib/src/services/text_input.dart:756]: Thanks for fixing my issue about a lack of docs here.  Some users had no idea what this was for.

**Review by justinmc — APPROVED** [review]
LGTM, assuming that no one is relying on the existing behavior for any good reason.  I can't think of why they would be.  And sorry it took me so long to get to this!

File: packages/flutter/lib/src/services/text_formatter.dart [hunk 3]
```diff
@@ -121,6 +121,94 @@ class _SimpleTextInputFormatter extends TextInputFormatter {
...
+  static _MutableTextRange? fromTextSelection(TextSelection selection) {
+    return selection.isValid
+      ? _MutableTextRange(selection.baseOffset, selection.extentOffset)
+      : null;
+  }
+
+  /// The start index of the range, inclusive.
+  ///
+  /// The value of [base] should always be greater than or equal to 0, and can
+  /// be larger than, smaller than, or equal to [extent].
```
Comment (chunhtai) [inline packages/flutter/lib/src/services/text_formatter.dart:143]: Do we need to worry about inverted selection?

File: packages/flutter/test/services/text_formatter_test.dart [hunk 4]
```diff
@@ -79,7 +80,7 @@ void main() {
       );
       expect(
         FilteringTextInputFormatter(RegExp('o+'), allow: false, replacementString: '*').formatEditUpdate(testOldValue, selectedIntoTheWoods),
-        const TextEditingValue(text: 'Int* the W**ds', selection: TextSelection(baseOffset: 11, extentOffset: 14)),
+        const TextEditingValue(text: 'Int* the W*ds', selection: TextSelection(baseOffset: 11, extentOffset: 13)),
```
Comment (chunhtai) [inline packages/flutter/test/services/text_formatter_test.dart:83]: Is this behavior change part of the fix?

File: packages/flutter/lib/src/services/text_formatter.dart (hunk 3 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/services/text_formatter.dart:143]: Oh right I think I just copied the documentation over. No I think the indices are processed separately and it doesn't care if base > extent or not. I'll add a test for this.

File: packages/flutter/lib/src/services/text_formatter.dart [hunk 5]
```diff
@@ -246,33 +327,81 @@ class FilteringTextInputFormatter extends TextInputFormatter {
   /// string) because both of the "o"s would be matched simultaneously
   /// by the pattern.
   ///
-  /// Additionally, each segment of the string before, during, and
-  /// after the current selection in the [TextEditingValue] is handled
-  /// separately. This means that, in the case of the "Into the Woods"
```
[the remark this answers is not in the document]
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/services/text_formatter.dart:251]: No idea. It's still a hard choice to make when an index is strictly within the replace range. The old implementation works in most cases I think.

File: packages/flutter/test/services/text_formatter_test.dart (hunk 4 above)
Comment (LongCatIsLooong) [reply packages/flutter/test/services/text_formatter_test.dart:83]: Yeah the double "o" was separated by the caret so the filtering was applied separately. Hopefully nobody is relying on this behavior.

File: packages/flutter/lib/src/services/text_formatter.dart (hunk 3 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/services/text_formatter.dart:143]: Oh nvm the documentation was right, it works fine with inverted selections. Many test cases have the selection range inverted.

**Review by chunhtai — APPROVED** [review]
LGTM

**Pull request merged**